### ISSUE TITLE
EpoxyTransport C++

### ISSUE THREAD

**Issue by alekseymt74** [Microsoft/bond#330]
We are working on distributed system and need the ability to be notified on server's activity changed: connection established / dropped, handshake status, etc.. It will help us to determine the status of connection.
It will be usefull to pass a callback to EpoxyTransport::Connect method like it's done in proxy's method call.

At this time I found the way to resolve a part (waiting for connection/handshake established) of this problem (see the code below). But it is too ugly and contains extra code and ambiguous issues..

````
  bond::comm::SocketAddress loopback( "127.0.0.1", 25188 );
  bond::comm::epoxy::EpoxyTransport transport;

  ::test::test_01::serv_01::Proxy::Using< std::promise > proxy( transport.Connect( loopback ) );
  std::promise< bond::comm::message< void > > promise;
  bool err = true;
 [*1] // without this delay between trying to establish connection and calling server's method my callback will never be called - need to be fixed
  // delay value is undetermined..
  std::this_thread::sleep_for( std::chrono::seconds( 5 ) );
  while( err )
  {
    // calling some fake method in order to determine server's accessibility
    proxy.check_alive( [ &err, &promise ]( const bond::comm::message< void > &response ) {
      promise.set_value( response );
      err = response.is_error();
    });
    promise.get_future().wait();
    if( err )
    {
      // wait for sometime to make next attempt
      std::this_thread::sleep_for( std::chrono::seconds( 1 ) );
      promise.swap( std::promise< bond::comm::message< void > >() );
    }
  }
  // connection established and server is ready for conversation
  //proxy.call_smth(..);
````

**Comment by alekseymt74** [comment]
I found the reason problem [ **1** ](see message above): 
file: **_epoxy.h_**, 
_**class: bond::comm::epoxy::detail::EpoxyTransport<WireProtocol,Allocator>::EpoxyConnection**_
method: **_Send(uint64_t id, Packet packet)_**
line: **_687, l->emplace_back(**0**, std::move(buffers));_**
It looks like misprint, :)
When we call remote method new conversation id is set in **_bond::comm::PacketTransport::ConnectionSink::StartCoversation_** and new listener with this id is added.
Later we come to **_EpoxyConnection::Send(uint64_t id, Packet packet)_** and if
**_EpoxyConnection::OnConnect_** was not already firered and therefore **_m_sending_** flag was not cleared buffer with **0(!)** conversation id is queued: **_l->emplace_back(0/***!***/, std::move(buffers));_**
So **_bond::comm::PacketTransport::ConnectionSink::CompleteConversation(uint64_t conversation_id)_** is called with wrong id and our callback cannot be fired.
If there is some delay **_EpoxyConnection::OnConnect_** fired, **_m_sending flag_** cleared, **_m_reconnect_** is set and this block of code in **_EpoxyConnection::Send(uint64_t id, Packet packet)_** is skipped and our callback is fired.
I changed **_l->emplace_back(0/***!***/, std::move(buffers))_** to **_l->emplace_back(id, std::move(buffers))_** and delay is not needed now.
P.S. I have not look farther ahead but probably this misprint can produce another problems

**Comment by alekseymt74** [comment]
I found the way to achieve the ability to be notified on connection status changes.

**Short description:**

I extended **EpoxyTransport** class and some of it inner classes:
**EpoxyConnection** - for handshake done notification,
**ClientConnection** - helper for **EpoxyConnection** extension. Also provides access to **EpoxyConnection::Connect** method
Made a wrapper for **PacketTransport::ConnectionSink** class to handle needed notifications: **ConnectionClosed()**, **ConnectionDropped()**. Adds **OnHandshakeDone()** method which is used from **EpoxyConnection** extension and tells that connection established.
**EpoxyTransport** extension takes a callback which handles notifications on connection status changes. On connection dropped i call **reconnect()** method of my transport extension, call some service method and wait for other notifications. If the connection would be temporary dropped i will receive this status and start to reconnect again waiting fo success.

**Questions:**
Certainly this code needs several iterations/refactoring/extension (i have many ideas on its extension both on client and server side). But the main problem - it tightly bound to **EpoxyTransport** structure. What parts of my code (see below) can be incorporated into base class (**EpoxyTransport**)?
If it's impossible what other approaches can i use?
In current version i will have to add my changes to bases on every new release.

**Thanks in advance!**

I tried to resolve it with minimum changes in base classes but could not completely avoid it.

**Code:**

These are modifications that were done in **epoxy.h** file:

`class EpoxyTransport::EpoxyConnection`
  moved to protected scope to extend in subclass
`bool EpoxyTransport::EpoxyConnection::HandleFrame(bool success, detail::ClassifyResult&& result)`
  moved to protected scope and made virtual to extend in subclass. Needed to determine successful handshake
```
void EpoxyTransport::EpoxyConnection::DoDisconnect()
{
  <..>
  // added to have the ability to recconect when temporary dropped connection happens
  if(m_handshake_received.load())
    BOOST_VERIFY(m_handshake_received.exchange(false));
}

```
```
void EpoxyTransport::EpoxyConnection::Send(uint64_t id, Packet packet) override
{
  <..>
    // replaced (see previous message)
    //l->emplace_back(0, std::move(buffers));
    l->emplace_back(id, std::move(buffers));
  <..>
}
```
`class EpoxyTransport`
```
// added virtual factory method to provide the ability to create extended class of EpoxyConnection
virtual boost::shared_ptr<EpoxyConnection> create_epoxy_connection(const Allocator& allocator,
																 const SocketAddress& address,
																 const boost::shared_ptr<INetworkConnectionSink>& handler,
																 const boost::shared_ptr<boost::asio::io_service>& ioServicePtr)
{
  return boost::make_shared<EpoxyConnection>(allocator, address, handler, ioServicePtr);
}

```
`class EpoxyTransport::ClientConnection`
  moved to protected scope to extend in subclass
`boost::shared_ptr<EpoxyConnection> EpoxyTransport::ClientConnection::m_connection;`
  moved to protected scope to be visible in subclass
`class EpoxyTransport::ClientConnection`
 ```
 // added constructor taking a factory which will create  the instance of EpoxyTransport::EpoxyConnection
  ClientConnection(const Allocator& allocator,
				   const SocketAddress& address,
				   const boost::shared_ptr<INetworkConnectionSink>& handler,
				   const boost::shared_ptr<ThreadService>& threadService
				   , std::function< boost::shared_ptr<EpoxyConnection>(const Allocator&, const SocketAddress&
				    , const boost::shared_ptr<INetworkConnectionSink>&, const boost::shared_ptr<boost::asio::io_service>&) > epoxy_conn_factory)
    : m_threadService(threadService)
    , m_connection(epoxy_conn_factory(allocator, address, handler, threadService->io_service()))
  {
    m_connection->Connect();
  }
```
`class EpoxyTransport`
```
// added virtual factory method to provide the ability to create extended class of ClientConnection. Also passes EpoxyConnection factory to its constructor
virtual boost::shared_ptr<ClientConnection> create_client_connection(const Allocator& allocator,
																	const SocketAddress& address,
																	const boost::shared_ptr<INetworkConnectionSink>& handler,
																	const boost::shared_ptr<ThreadService>& threadService
																  , std::function< boost::shared_ptr<EpoxyConnection>(const Allocator&, const SocketAddress&
																	, const boost::shared_ptr<INetworkConnectionSink>&, const boost::shared_ptr<boost::asio::io_service>&) > epoxy_conn_factory)
{
  return boost::make_shared<ClientConnection>(allocator, address, handler, threadService, epoxy_conn_factory);
}
```
```
class EpoxyTransport
  boost::shared_ptr<INetworkConnection>
  StartClientConnection(const SocketAddress& address,
					    const boost::shared_ptr<INetworkConnectionSink>& handler) override
  {
	// uses ClientConnection factory method
	return create_client_connection(m_allocator, address, handler, m_threadService, std::bind( &EpoxyTransport::create_epoxy_connection, this
	  , std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, std::placeholders::_4));
  }

```

**My code:**
**Extensions:**

```
namespace bond { namespace comm
{ 
  namespace epoxy { namespace detail
  { 
    namespace ext
    {
      // extending bond::comm::epoxy::detail::EpoxyTransport<WireProtocol, Allocator>
      template <typename WireProtocol,
      typename Allocator = std::allocator<char>>
      class EpoxyTransport : public bond::comm::epoxy::detail::EpoxyTransport<WireProtocol, Allocator>
      {
      public:
        enum EConnectionStatus { ConnClosed, ConnDropped, ConnEstablished };
        typedef std::function<void(EConnectionStatus, EpoxyTransport &)> conn_status_changed_cb;
      private:
        typedef typename PacketTransport<SocketAddress, WireProtocol>::INetworkConnection INetworkConnection;
        typedef typename PacketTransport<SocketAddress, WireProtocol>::INetworkConnectionSink INetworkConnectionSink;
        // wrapper needed to handle notifications. Also also handles handshake notification
        struct ConnectionSinkExt : public PacketTransport< SocketAddress, WireProtocol >::INetworkConnectionSink
        {
          // provided callback for handle notifications on connection status changes
          conn_status_changed_cb conn_st_cb;
          // transport, the owner this wrapper. Will be passed to callback
          EpoxyTransport *transport;
          // wrapped handler. All handled notifications will be delegated to it
          boost::shared_ptr<INetworkConnectionSink> handler;
          ConnectionSinkExt(boost::shared_ptr<INetworkConnectionSink> _handler, conn_status_changed_cb _conn_st_cb, EpoxyTransport *_transport)
            : handler(_handler), conn_st_cb( _conn_st_cb), transport(_transport){}
          void ConnectionClosed() override 
          {
            handler->ConnectionClosed();
            conn_st_cb(ConnClosed, *transport);
          }
          void ConnectionDropped() override
          {
            handler->ConnectionDropped();
            conn_st_cb(ConnDropped, *transport);
          }
          void PacketReceived(bool success, uint64_t conversation_id, Packet packet) override
          {
            handler->PacketReceived(success, conversation_id, packet);
          }
          void PacketSent(bool success, uint64_t conversation_id) override
          {
            handler->PacketSent(success, conversation_id);
          }
          virtual void OnHandshakeDone()
          {
            conn_st_cb(ConnEstablished, *transport);
          }
        };
        // ClientConnection extension. Need to provide required EpoxyConnection factory to the base class and access to EpoxyTransport::Connect method
        class ClientConnectionExt : public ClientConnection
        {
        public:
          ClientConnectionExt(const Allocator& allocator,
            const SocketAddress& address,
            const boost::shared_ptr<ConnectionSinkExt>& handler,
            const boost::shared_ptr<ThreadService>& threadService
            , std::function< boost::shared_ptr<EpoxyConnection>(const Allocator&, const SocketAddress&
              , const boost::shared_ptr<INetworkConnectionSink>&, const boost::shared_ptr<boost::asio::io_service>&) > epoxy_conn_factory)
            : ClientConnection( allocator, address, handler, threadService, epoxy_conn_factory )
          {}
          void Connect() { m_connection->Connect(); }
        };
        // EpoxyConnection extension. Need to reimplement EpoxyConnection::HandleFrame
        class EpoxyConnectionExt : public EpoxyConnection
        {
          // our wrapper which will handle handshake notification
          boost::shared_ptr<ConnectionSinkExt> handler;
        public:
          EpoxyConnectionExt(const Allocator& allocator,
                              const SocketAddress& address,
                              const boost::shared_ptr<ConnectionSinkExt>& _handler,
                              const boost::shared_ptr<boost::asio::io_service>& ioServicePtr)
            : EpoxyConnection( allocator, address, _handler, ioServicePtr)
            , handler( _handler )
          {}
          EpoxyConnectionExt(const Allocator& allocator,
                              const SocketAddress& address,
                              boost::asio::ip::tcp::socket&& socket,
                              const boost::shared_ptr<boost::asio::io_service>& ioServicePtr)
            : EpoxyConnection( allocator, address, socket, ioServicePtr)
          {}
          virtual bool HandleFrame(bool success, detail::ClassifyResult&& result) override
          {
            bool ret = EpoxyConnection::HandleFrame(success, std::forward<detail::ClassifyResult>(result));
            // Check successful handshake
            if (success && detail::FrameDisposition::ProcessConfig == result.disposition && ret)
              handler->OnHandshakeDone();
            return ret;
          }
        };
        const SocketAddress *address;
        boost::shared_ptr< ClientConnectionExt > cl_conn;
        conn_status_changed_cb conn_st_cb;
      protected:
        // reimplementation of ClientConnection factory
        virtual boost::shared_ptr<ClientConnection> create_client_connection(const Allocator& allocator,
          const SocketAddress& _address,
          const boost::shared_ptr<INetworkConnectionSink>& handler,
          const boost::shared_ptr<ThreadService>& threadService
          , std::function< boost::shared_ptr<EpoxyConnection>(const Allocator&, const SocketAddress&
            , const boost::shared_ptr<INetworkConnectionSink>&, const boost::shared_ptr<boost::asio::io_service>&) > epoxy_conn_factory) override
        {
          address = &_address;
          cl_conn = boost::make_shared<ClientConnectionExt>(allocator, _address, boost::make_shared<ConnectionSinkExt>(handler, conn_st_cb, this), threadService, epoxy_conn_factory);

          return cl_conn;
        }
        // reimplementation of EpoxyConnection factory
        virtual boost::shared_ptr<EpoxyConnection> create_epoxy_connection(const Allocator& allocator,
          const SocketAddress& address,
          const boost::shared_ptr<INetworkConnectionSink>& handler,
          const boost::shared_ptr<boost::asio::io_service>& ioServicePtr) override
        {
          return boost::make_shared<EpoxyConnectionExt>(allocator, address, boost::dynamic_pointer_cast<ConnectionSinkExt, INetworkConnectionSink>(handler), ioServicePtr);
        }
      public:
        EpoxyTransport(conn_status_changed_cb _conn_st_cb)
          : conn_st_cb( _conn_st_cb )
        {}
        const SocketAddress &get_address() const { return *address; }
        // reconnect on dropped connection
        void reconnect()
        {
          std::thread([this]()
          {
            std::this_thread::sleep_for(std::chrono::seconds(1));
            cl_conn->Connect();
          }).detach();
        }
      };
    } // ext
  } // epoxy::detail
  namespace ext
  {
    // declaration of EpoxyTransport extension
    typedef LayerTransport<
      detail::ext::EpoxyTransport<CompactWireProtocol>, // transport impl
      CompactWireProtocol // wire protocol to be used by layers
    > EpoxyTransport;
  } // epoxy::ext
} } } // bond::comm::epoxy

```

**TEST:**

```
  bond::comm::SocketAddress loopback( "127.0.0.1", 25188 );
  std::promise< bool > conn_set_promise;
  bond::comm::epoxy::ext::EpoxyTransport transport([&conn_set_promise](bond::comm::epoxy::ext::EpoxyTransport::EConnectionStatus conn_status
    , bond::comm::epoxy::detail::ext::EpoxyTransport< bond::comm::CompactWireProtocol > &transport)
  {
    switch (conn_status)
    {
    case bond::comm::epoxy::ext::EpoxyTransport::EConnectionStatus::ConnClosed:
      std::cout << "connection to " << transport.get_address() << " closed" << std::endl;
      break;
    case bond::comm::epoxy::ext::EpoxyTransport::EConnectionStatus::ConnDropped:
      std::cout << "connection to " << transport.get_address() << " dropped" << std::endl;
      // trying to reconnect
      transport.reconnect();
      break;
    case bond::comm::epoxy::ext::EpoxyTransport::EConnectionStatus::ConnEstablished:
      std::cout << "connection on " << transport.get_address() << " established" << std::endl;
      // connection established - notify std::future waiting on it
      conn_set_promise.set_value(true);
      break;
    }
  });
  ::test::test_01::serv_01::Proxy::Using< std::promise > proxy(transport.Connect(loopback));
  std::thread([&conn_set_promise, &proxy]()
  {
    while (true)
    {
      conn_set_promise.get_future().wait();
      conn_set_promise.swap(std::promise< bool >());
      // connection established. Call some method and wait for another one (if we will manually drop/start server)
      proxy.check_alive();
    }
  }).detach();
```

**Comment by chadwalters** [comment]
Aleksey, thanks for providing the above code. We definitely want to the connect/disconnect notification feature in C++ for symmetry with the C# implementation. Can you submit this as a pull request and sign the CLA? That is the usual mechanism that we use for reviewing code contributions. Thanks!

**Comment by alekseymt74** [comment]
Hi, Chad!
I'll try to make a pull request next week. I made several implementations of existing classes and put them in separate namespace in order to keep compability with all previous test and existing projects. Also some insignificant changes of existing code were made. I will provide detailed explanation in commit messages and code comments.

**Comment by alekseymt74** [comment]
I'm ready to submit pull request.
I'm already signed CLA and received confirmation.
What is the next step?

**Comment by chwarr** [comment]
The next step would be to start the pull request. There are more details in [CONTRIBUTING.md](https://github.com/Microsoft/bond/blob/master/CONTRIBUTING.md#process). Let us know if you have questions.

**Comment by alekseymt74** [comment]
I tried to push my local branch to origin: "git push -u origin cpp-comm-conn-state-layer". But received: "
remote: Permission to Microsoft/bond.git denied to alekseymt74.
fatal: unable to access 'https://github.com/Microsoft/bond.git/': The requested URL returned error: 403"
What i'm doing wrong?

**Comment by alekseymt74** [comment]
... I see my mistake: i forgot to fork and trying to work with origin.. I will fix it

**Comment by chwarr** [comment]
Here's the PR: https://github.com/Microsoft/bond/pull/350

**Comment by chwarr** [comment]
Since Bond Comm is now deprecated, we aren't planning to addresses this.
